REP-2001: Desktop-full, perception and simulation variants for ROS 2 (#334)
ros-infrastructure/rep: rep-2001.rst

End-user entry points
---------------------

We define four main entry points for ROS users.

>  * `desktop_full` (recommended from Humble): The `desktop_full` metapackage is the main entry point for users.

Reviewer: My preference is not to have a `desktop_full` variant at all.  I have two main reasons for this:

1. The name `desktop_full` is really meaningless; it gives no indication of what is in there, or what is expected to be in there.
2. Because it is so ambiguous, it means that it can be continually argued over what should and shouldn't be in it.

Author: I hear you. Here are some other suggestions:

1. quickstart
2. entrypoint
3. toolkit

Reviewer 2: I agree the name isn't optimal. We could do `kitchen_sink`?? just kidding. We are putting it all in there. We used to have a automagic `all` metapackage. Which actually depended on all built packages, but that was deprecated a long time ago as being truely meaningless as to what you actually got as it was best effort. And you could just replace it with `ros-XXX-*` in apt.

I like the concept of `quickstart` where it's indicating that this is a shorthand guess at what you might want but isn't necessarily the end all of where you should end up for an application. We want to capture the idea of it being a default starting point but you don't have to stay with it once you know better. Looking at the [ubuntu metapackages](https://help.ubuntu.com/community/MetaPackages) they have a concept of `standard` Which by name is nice, but I think its closer to `ros-base` semantically if we want to keep parallels. 

`entrypoint` is a nice idea, but already is heavily overloaded with other meanings as technical keyword. 

`toolkit` has a similar problem with `desktop_full` as it doesn't give  a sense of scope.

So in the end, I'd suggest `quickstart` is best so far.

Author: Changed to `quickstart` in 86241d49b523304ce7d95f877f2e7cb38a5ace70.

Author: After a discussion with the ROS team, we'll go back to `desktop_full` for backwards compatibility with ROS 1: [c0dd46b](https://github.com/ros-infrastructure/rep/pull/334/commits/c0dd46bba6e7e24c949ee72a9e3174ee167bd958)

End-user entry points
---------------------

We define various entry points for ROS users.

>  * `desktop_full` (recommended from Humble): The `desktop_full` variant is the main entry point for new users.
>    It provides a "batteries included" experience, enabling novice users to complete most entry
>    tutorials without knowledge of the underlying library structure.
>  * `desktop` (recommended until Galactic): The `desktop` variant provides all commonly used libraries as well as

Reviewer: So given what we discussed earlier, I think we should change the wording here, and in particular don't say that `desktop_full` is recommended from Humble.  That's because we aren't guaranteeing that we are going to keep it available; if one of the packages in it doesn't work, we can proceed anyway.  So I'd change this to:

```suggestion
 * `desktop_full` (available from Humble): The `desktop_full` variant provides a "batteries included" experience, enabling novice users to complete most entry tutorials without knowledge of the underlying library structure.
 * `desktop` (recommended): The `desktop` variant provides all commonly used libraries as well as
```

Author: Fair enough, commited

                 teleop_twist_keyboard, tlsf, tlsf_cpp,
                 topic_monitor, turtlesim]

Perception
""""""""""

::

  - perception:
      extends:  [ros_base]
      packages: [image_common, image_pipeline, image_transport_plugins,
                 laser_filters, laser_geometry, perception_pcl,
                 vision_opencv]

Simulation
""""""""""

::

  - simulation:

>       extends:  [ros_base]

Reviewer: I'd actually suggest that the `simulation` variant extend `desktop`.  Presumably if a user is going to use simulation, they also want general debugging and other GUI tools.  But I don't feel strongly about it, so if you think `ros_base` is more appropriate I'm fine with it.

Author: I don't have a strong feeling either way, so I roughly followed what was done in ROS 1, where it extends `robot`.

Reviewer 2: I would suggest keeping with the base. So that the simulation can be used in a more embedded type of environment, such as on a cloud machine. There will be some integrated unused gui elements but this sort of application. But the developer tools and introspection tools are not necessary. And that union of desktop + simulation is covered effectively desktop_full with only perception added in addition.

Reviewer 3: > embedded type of environment

headless environment, probably the better way to say it. And yeah, we run lots of headless simulations in RoboMaker, for example. The gazebo frontend is never invoked in the actual simulation environment, and instead can be connected at will from another instance.

Author: So, `ros_ign` includes `ros_ign_gazebo_demos`, which [brings tools like RQt and RViz](https://github.com/ignitionrobotics/ros_ign/blob/8530ab1189d0738a2af438ff8d740681abf794ec/ros_ign_gazebo_demos/package.xml#L17-L25). If we're targeting a headless use case, I'd break this up a bit more and remove the demos from `simulation`, leaving them only on the `desktop-full` / `quickstart` variant, which builds on top of `desktop`.

Author: Removed `ros_ign_gazebo_demos` from the `simulation` variant in 86241d49b523304ce7d95f877f2e7cb38a5ace70.
